esp/esp/program/modules/handlers/studentregphasezeromanage.py

```diff
@@ -0,0 +1,161 @@
+  (see AUTHORS file)
+
+The ESP Web Site is free software; you can redistribute it and/or
+modify it under the terms of the GNU Affero General Public License
+as published by the Free Software Foundation; either version 3
+of the License, or (at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public
+License along with this program; if not, write to the Free Software
+Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+
+Contact information:
+MIT Educational Studies Program
+  84 Massachusetts Ave W20-467, Cambridge, MA 02139
+  Phone: 617-253-4882
+  Email: esp-webmasters@mit.edu
+Learning Unlimited, Inc.
+  527 Franklin St, Cambridge, MA 02139
+  Phone: 617-379-0178
+  Email: web-team@learningu.org
+"""
+
+from esp.utils.web import render_to_response
+from esp.middleware.threadlocalrequest import get_current_request
+from esp.program.modules.base import ProgramModuleObj, main_call, aux_call, meets_deadline, needs_student, meets_grade, meets_cap, no_auth, needs_admin
+from esp.users.models import Record, ESPUser, Permission
+from esp.program.models import PhaseZeroRecord
+from esp.program.modules.forms.phasezero import SubmitForm
+from esp.tagdict.models import Tag
+
+from django.contrib.auth.models import Group
+from django.db.models.query import Q
+
+import copy, datetime, json
+
+class StudentRegPhaseZeroManage(ProgramModuleObj):
+    def isCompleted(self):
+        return get_current_request().user.canPassPhaseZero(self.program)
+
+    @classmethod
+    def module_properties(cls):
+        return {
+            "module_type": "manage",
+            'required': False,
+            'admin_title': 'Manage Student Registration Phase Zero',
+            'link_title': 'Manage Phase Zero',
+        }
+
+    def lottery(self, prog, role):
+        # Run lottery algorithm.
+        # Get grade caps
+        grade_caps_str = prog._grade_caps()
+        grade_caps = {int(key[0]):grade_caps_str[key] for key in grade_caps_str}
+
+        #Get lottery records and randomize them
+        records = PhaseZeroRecord.objects.filter(program=prog).order_by('?')
+
+        ###############################################################################
+        # The lottery algorithm is run, with randomization and processing in order.
+        # If any one in the group doesn't get in (due to cap size), no one in that group gets in.
+        counts = {key:0 for key in grade_caps}
+        winners = Group.objects.get_or_create(name=role)[0]
+
+        for i in records:
+            sibs = i.user.all()
+            newcounts = copy.copy(counts)
+            for j in sibs:
+                newcounts[j.getGrade(prog)] += 1
+
+            cpass = True
```

Comment:
Probably easier and clearer to do `cpass = not any(newcounts[c] > grade_caps[c] for c in counts)`